### PR TITLE
Support `PermMem` metadata section and user-defined metadata

### PR DESCRIPTION
**Pull request by gaojiaqi7**
Align the implementation with https://github.com/confidential-containers/td-shim/blob/main/doc/tdshim_spec.md#metadata-definition

### CODE REVIEW

**Comment by gaojiaqi7** [comment]
Fix https://github.com/confidential-containers/td-shim/issues/449

File: td-shim/src/bin/td-shim/info.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+impl<'a> ShimInfo<'a> {
+    // Validate the metadata and get the basic infomation from
+    // it if any
+    pub fn new() -> Option<Self> {
+        let metadata_offset = unsafe { *((u32::MAX - TDX_METADATA_OFFSET + 1) as *const u32) };
+        if metadata_offset >= TD_SHIM_FIRMWARE_SIZE
+            || metadata_offset < size_of::<Guid>() as u32
+            || metadata_offset > TD_SHIM_FIRMWARE_SIZE - size_of::<TdxMetadataDescriptor>() as u32
+        {
+            error!("Invalid TDX Metadata offset\n");
```
Comment (jyao1) [inline td-shim/src/bin/td-shim/info.rs:43]: I remember we support both OFFSET style and GUID style. Should we support both cases here?

File: td-shim/src/bin/td-shim/info.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+
+use alloc::vec::Vec;
```
Comment (jyao1) [inline td-shim/src/bin/td-shim/info.rs:5]: Can we rename file to be shim_info?

File: td-shim/src/bin/td-shim/info.rs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    // If metadata contains the `TD_HOB` section, TD-Shim
+    // can get additional information from TD HOB
+    pub td_hob: Option<&'static [u8]>,
+
+    // If metadata contains one/more `PermMem` sections,
+    // TD-Shim should ignore the memory information in TD HOB.
+    metadata_has_perm: bool,
+
+    pub memory: Vec<ResourceDescription>,
+    pub acpi_tables: Vec<&'a [u8]>,
```
Comment (jyao1) [inline td-shim/src/bin/td-shim/info.rs:29]: I feel it is weird to put acpi_tables here.

I assume ShimInfo means the metadata / memory related static information, right?

Or you want to create a structure to hold Runtime information ?

File: td-shim/src/bin/td-shim/info.rs (hunk 1 above)
Comment (gaojiaqi7) [reply td-shim/src/bin/td-shim/info.rs:43]: There is always an `OFFSET` at `4GiB - 0x20` no matter if there is a GUID chain or not

File: td-shim/src/bin/td-shim/info.rs (hunk 2 above)
Comment (gaojiaqi7) [reply td-shim/src/bin/td-shim/info.rs:5]: Sure

File: td-shim/src/bin/td-shim/info.rs (hunk 3 above)
Comment (gaojiaqi7) [reply td-shim/src/bin/td-shim/info.rs:29]: I agree that we should separate the different kind of informations. The `ShimInfo` is just derived from original `TdHobInfo` structure and combined `1` and `2`

**Review by jyao1 — APPROVED** [review]

**Pull request merged**